Include/cpython/pyatomic.h

```diff
@@ -0,0 +1,397 @@
+#ifndef Py_ATOMIC_H
+#define Py_ATOMIC_H
+
+// This header provides cross-platform low-level atomic operations
+// similar to C11 atomics.
+//
+// Operations are sequentially consistent unless they have a suffix indicating
+// otherwise. If in doubt, prefer the sequentially consistent operations.
+//
+// The "_relaxed" suffix for load and store operations indicates the "relaxed"
+// memory order. They don't provide synchronization, but (roughly speaking)
+// guarantee somewhat sane behavior for races instead of undefined behavior.
+// In practice, they correspond to "normal" hardware load and store instructions,
+// so they are almost as inexpensive as plain loads and stores in C.
+//
+// Note that atomic read-modify-write operations like _Py_atomic_add_* return
+// the previous value of the atomic variable, not the new value.
+//
+// See https://en.cppreference.com/w/c/atomic for more information on C11 atomics.
+// See https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p2055r0.pdf
+// "A Relaxed Guide to memory_order_relaxed" for discussion of and common usage
+// or relaxed atomics.
+
+// Atomically adds `value` to `address` and returns the previous value
+static inline int
+_Py_atomic_add_int(int *address, int value);
+
+static inline int8_t
+_Py_atomic_add_int8(int8_t *address, int8_t value);
+
+static inline int16_t
+_Py_atomic_add_int16(int16_t *address, int16_t value);
+
+static inline int32_t
+_Py_atomic_add_int32(int32_t *address, int32_t value);
+
+static inline int64_t
+_Py_atomic_add_int64(int64_t *address, int64_t value);
+
+static inline intptr_t
+_Py_atomic_add_intptr(intptr_t *address, intptr_t value);
+
+static inline unsigned int
+_Py_atomic_add_uint(unsigned int *address, unsigned int value);
+
+static inline uint8_t
+_Py_atomic_add_uint8(uint8_t *address, uint8_t value);
+
+static inline uint16_t
+_Py_atomic_add_uint16(uint16_t *address, uint16_t value);
+
+static inline uint32_t
+_Py_atomic_add_uint32(uint32_t *address, uint32_t value);
+
+static inline uint64_t
+_Py_atomic_add_uint64(uint64_t *address, uint64_t value);
+
+static inline uintptr_t
+_Py_atomic_add_uintptr(uintptr_t *address, uintptr_t value);
+
+static inline Py_ssize_t
+_Py_atomic_add_ssize(Py_ssize_t *address, Py_ssize_t value);
+
+// Performs an atomic compare-and-exchange. If `*address` and `expected` are equal,
+// then `value` is stored in `*address`. Returns 1 on success and 0 on failure.
+// These correspond to the "strong" variations of the C11 atomic_compare_exchange_* functions.
```

Comment:
C11 passes `expected` as a pointer, so that it's updated with the actual value when the latter doesn't match the former. Why not keep that convention here?

Reply:
The motivation was two-fold: First, many of the _Py_atomic_compare_exchange calls in the nogil fork use constants as `expected` and this would be more verbose if it needs to be a pointer, e.g.:

```
_Py_atomic_compare_exchange_uint8(&m->v, LOCKED, UNLOCKED)
```

vs.

```
uint8_t expected = LOCKED:
_Py_atomic_compare_exchange_uint8(&m->v, &expected, UNLOCKED)
```

Second, I find this style (no pointer for expected) to be a bit less error-prone. I've been tripped up once or twice by having `expected` be modified when I didn't expect it.

I don't feel terribly strongly about this, so if there is a general preference for sticking closer to the C11-style API here, I can change it.

Reply:
AFAIU the main motivation for the C11 style APIs is for retry loops in lockless data structure implementations. A simplistic example (this may be embarassingly wrong):
```c

struct ListNode;

typedef struct ListNode {
  int value;
  struct ListNode* next;
} ListNode;

void ListAppend(ListNode* list, int new_value) {
  ListNode* new_node = (ListNode*) malloc(sizeof ListNode);
  new_node->value = new_value;
  new_node->next = NULL;
  ListNode* expected = NULL;
  while (_Py_atomic_compare_exchange_ptr(&list->next, &expected, new_node)) {
    list = expected;
    expected = NULL;
  }
}
```

Reply:
If the second argument is a reference (``&expected``), does it mean that it changes the value of the second argument (``*expected``)?

For C11 atomic_exchange(), the second argument is not a pointer, but a value (integer), no? https://en.cppreference.com/w/c/atomic/atomic_exchange 

C11 atomic_compare_exchange_strong() and atomic_compare_exchange_weak() use a pointer for expected. But this API writes into ``*expected`` if the ``*obj`` is not equal to ``*expected``.

> The behavior of atomic_compare_exchange_* family is as if the following was executed atomically:

```c
if (memcmp(obj, expected, sizeof *obj) == 0) {
    memcpy(obj, &desired, sizeof *obj);
    return true;
} else {
    memcpy(expected, obj, sizeof *obj);
    return false;
}
```

For this header fie, I would prefer to not have two flavors, the API is already quite long! I would prefer to have a single flavor. If there is an usecase where setting expected is relevant, I suggest to use a pointer for the second argument.

In short, I agree to change the API to ``int _Py_atomic_compare_exchange_int32(int32_t *obj, int32_t *expected, int32_t desired)``. The behavior should be well documented.

*obj*, *expected* and *desired* names come from the C11 API.

Reply:
I'll make the second argument a pointer like the C11 API.